pkg/cmd/dev/ui.go

```diff
@@ -0,0 +1,192 @@
+	// clusterUI is the absolute path to ./pkg/ui/workspaces/cluster-ui
+	clusterUI string
+	// dbConsole is the absolute path to ./pkg/ui/workspaces/db-console
+	dbConsole string
+}
+
+// getUIDirs computes the absolute path to the root of each UI sub-project.
+func getUIDirs(d *dev) (*UIDirectories, error) {
+	workspace, err := d.getWorkspace(d.cli.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	return &UIDirectories{
+		clusterUI: path.Join(workspace, "./pkg/ui/workspaces/cluster-ui"),
+		dbConsole: path.Join(workspace, "./pkg/ui/workspaces/db-console"),
+	}, nil
+}
+
+// makeUIWatchCmd initializes the 'ui watch' subcommand, which sets up a live-reloading HTTP server for db-console and a
+// file-watching rebuilder for cluster-ui.
+func makeUIWatchCmd(d *dev) *cobra.Command {
+	const (
+		// portFlag is the name of the long (GNU-style) flag that controls which port webpack's dev server listens on
+		portFlag = "port"
+		// dbTargetFlag is the name of the long (GNU-style) flag that determines which DB instance to proxy to
+		dbTargetFlag = "db"
+		// secureFlag is the name of the boolean long (GNU-style) flag that makes webpack's dev server use HTTPS
+		secureFlag = "secure"
+	)
+
+	watchCmd := &cobra.Command{
+		Use:   "watch",
+		Short: "Watches for file changes and automatically rebuilds",
+		Long: `Starts a local web server that watches for JS file changes, automatically regenerating the UI
+
+Replaces 'make ui-watch'.`,
+		Args: cobra.MinimumNArgs(0),
+		RunE: func(cmd *cobra.Command, commandLine []string) error {
+			// Create a context that cancels when OS signals come in
+			ctx, stop := signal.NotifyContext(d.cli.Context(), os.Interrupt, os.Kill)
+			defer stop()
+
+			isOss, err := cmd.Flags().GetBool(ossFlag)
+			if err != nil {
+				return err
+			}
+
+			// Build prerequisites for db-console and cluster-ui
+			args := []string{
+				"build",
+				"//pkg/ui/workspaces/db-console/src/js:crdb-protobuf-client",
+			}
+			if !isOss {
+				args = append(args, "//pkg/ui/workspaces/db-console/ccl/src/js:crdb-protobuf-client-ccl")
+			}
+			logCommand("bazel", args...)
+			err = d.exec.CommandContextInheritingStdStreams(ctx, "bazel", args...)
+
+			if err != nil {
+				log.Fatalf("failed to build UI watch prerequisites: %v", err)
+				return err
+			}
+
+			// Extract remaining flags
+			portNumber, err := cmd.Flags().GetInt16(portFlag)
+			if err != nil {
+				log.Fatalf("unexpected error: %v", err)
+				return err
+			}
+			port := fmt.Sprint(portNumber)
+			dbTarget, err := cmd.Flags().GetString(dbTargetFlag)
+			if err != nil {
+				log.Fatalf("unexpected error: %v", err)
+				return err
+			}
+			_, err = url.Parse(dbTarget)
+			if err != nil {
+				log.Fatalf("invalid format for --%s argument: %v", dbTargetFlag, err)
+				return err
+			}
+			secure := mustGetFlagBool(cmd, secureFlag)
+
+			// `yarn` is required to be on a user's path, since it won't run via Bazel
+			err = d.ensureBinaryInPath("yarn")
```

Comment:
This is fine for now (I'm sure UI developers all have `yarn` installed anyway), but it's actually not required, since `yarn` is provided as `@nodejs//:yarn`. For my machine it's `/private/var/tmp/_bazel_ricky/be70b24e7357091e16c49d70921b7985/execroot/cockroach/external/nodejs_darwin_arm64/bin/yarn`. At this point I don't have a sensible way to get the path to that `yarn` though, so you can ignore this for now.

Reply:
I was looking for something like that last night!  I couldn't find anything short of building a platform-dependent path off of `bazel info execution_root`, which is kind of gnarly 😞